In [2]:
import numpy as np

I seperated this problem into two parts
1. find the expected number of steps on the football
2. find the probability, that the number of steps is less or equal on the kitchen floor.
The inverse of this probability is the value we are looking for

## 1. Step: get the expected number of steps on the ball

The ball can be represented by a graph of pentagons.
if you mark a random node x with 0 and every other node with the shortest distance to x you observe the following pattern:

node:  neighbours:
0      1,1,1
1      0,2,2
2      1,2,3
3      2,3,4
4      3,3,5
5      4,4,4

The graph can then be collapsed to the graph with the following adjazenzy matrix:

In [3]:
A =    [[0  , 1  , 0  , 0  , 0  , 0 ],
        [1/3, 0  , 2/3, 0  , 0  , 0  ],
        [0  , 1/3, 1/3, 1/3, 0  , 0  ],
        [0  , 0  , 1/3, 1/3, 1/3, 0  ],
        [0  , 0  , 0  , 2/3, 0  , 1/3],
        [0  , 0  , 0  , 0  , 1  , 0  ]]

Let $p_{i,j}$ be the probability to transition from state i to state j, which corresponds to the the $(i,j)^th$ element in the adjazenzy matrix.

Let $E_i$ be the expected number of steps to reach an absorbing state from state i. (absorbing state = state that has no outgoing edges) 

Then $E_i$ must satisfy the following equation:

$E_i = 1 + p_{i,1}E_1 + p_{i,2}E_2 + ... + p_{i,n}E_n = 1 + \sum_{j = 1}^{n} p_{i,j}E_j$

$E_i$ corresponds to the dot product of the Expected Value Vector and the i-th row of the adjazenzy matrix + 1.

We can derive the following formula:

We still have the Problem, that 0 is no absorbing state. Thus we delete all outgoing edges to other states:

In [4]:
A =    [[1  , 0  , 0  , 0  , 0  , 0 ],
        [1/3, 0  , 2/3, 0  , 0  , 0  ],
        [0  , 1/3, 1/3, 1/3, 0  , 0  ],
        [0  , 0  , 1/3, 1/3, 1/3, 0  ],
        [0  , 0  , 0  , 2/3, 0  , 1/3],
        [0  , 0  , 0  , 0  , 1  , 0  ]]

If we look closely we see that I-A will lead to the first row only containing zeros which makes the equation system not solvable. To compensate for this we remove the first row and first column, solve for $E_1$ and because we will transition from state 0 to state 1 with 100% we get: $E_0 = 1 + E_1$

In [5]:
A =    [[0  , 2/3, 0  , 0  , 0  ],
        [1/3, 1/3, 1/3, 0  , 0  ],
        [0  , 1/3, 1/3, 1/3, 0  ],
        [0  , 0  , 2/3, 0  , 1/3],
        [0  , 0  , 0  , 1  , 0  ]]
A = np.matrix(A)

In [6]:
I =    [[1 , 0 , 0 , 0 ,0],
        [0 , 1 , 0 , 0 ,0],
        [0 , 0 , 1 , 0 ,0],
        [0 , 0 , 0 , 1 ,0],
        [0 , 0 , 0 , 0 ,1]]
I = np.matrix(I)

In [7]:
one = [[1],[1],[1],[1],[1]]
one = np.matrix(one)

In [8]:
E = np.linalg.solve(I-A, one)
E

matrix([[19.],
        [27.],
        [32.],
        [34.],
        [35.]])

In [9]:
E_0 = E[0] +1
print("The expected number of steps starting at position 0 is: " + str(E_0))

The expected number of steps starting at position 0 is: [[20.]]


## 2. Step 2: Porobability of walking more than 20 steps on the kitchen floor

In [10]:
from fractions import Fraction

In [11]:
onethird = Fraction(1)/Fraction(3)
twothird = Fraction(2)/Fraction(3)

def step(state, steps, prob, path_probs , path):
    path += str(state)+ ", "
    
    if(state > steps):
        return;
    
    if(state == steps):
        if(state == 1):
            l.append(prob*onethird)
            #print(path + "0 | " + str(prob*onethird))
            
        elif (state%2 == 0):
            step(state-1, steps-1, prob*onethird, path_probs, path)
        
        else:
            step(state-1, steps-1, prob*twothird, path_probs, path)
    else:
        if(state == 0):
            step(1, steps-1, prob, path_probs, path)
        
        elif(state == 1):
            step(state+1, steps-1, prob*twothird, path_probs, path)
            
        elif(state % 2 == 0):
            step(state -1, steps-1, prob*onethird, path_probs, path)
            step(state +1, steps-1, prob*twothird, path_probs, path)
        else:
            step(state -1, steps-1, prob*twothird, path_probs, path)
            step(state +1, steps-1, prob*onethird, path_probs, path)
            
    

In [12]:
l = []

for i in range(1,21):
    #print("steps: " + str(i))
    step(0, i, Fraction(1), l,  "")

s = Fraction(0)
for p in l:
    s += p

probability = Fraction(1) - s

print(probability)
print(float(probability))



451116820/1162261467
0.3881371213006066


With a probability of 38.81371% the ant will take more steps on the kitchen floor